pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -418,23 +424,76 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
 		return types.InstantVectorSeriesData{}, nil
 	}
 
-	slices.SortFunc(data, func(a, b types.InstantVectorSeriesData) int {
-		return int(a.Floats[0].T - b.Floats[0].T)
-	})
+	// Merge floats and histograms individually.
+	// After which we check if there are any duplicate points in either the floats or histograms.
+
+	ds := dataSeries{
+		data:                 data,
+		sourceSeriesIndices:  sourceSeriesIndices,
+		sourceSeriesMetadata: sourceSeriesMetadata,
+	}
+
+	floats, err := b.mergeOneSideFloats(ds, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+	histograms, err := b.mergeOneSideHistograms(ds, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
 
-	mergedSize := len(data[0].Floats)
+	// Check for any conflicts between floats and histograms
+	i, j := 0, 0
+	for i < len(floats) && j < len(histograms) {
+		if floats[i].T == histograms[j].T {
+			// Conflict found
+			return types.InstantVectorSeriesData{}, fmt.Errorf("found both float and histogram samples for the match group FIXME on the %s side of the operation at timestamp %s", side, timestamp.Time(floats[i].T))
+		}
+		if floats[i].T < histograms[j].T {
+			i++
+		} else {
+			j++
+		}
+	}
+
+	return types.InstantVectorSeriesData{Floats: floats, Histograms: histograms}, nil
+}
+
+func (b *BinaryOperation) mergeOneSideFloats(ds dataSeries, side string) ([]promql.FPoint, error) {
+	if len(ds.data) == 0 {
+		return nil, nil
+	}
+	if len(ds.data) == 1 {
+		return ds.data[0].Floats, nil
+	}
+
+	ds.sortByFloat = true
+	sort.Sort(ds)
+
+	// After sorting, if the first series has no floats, we're done, as that means there are no floats in any series
+	if len(ds.data[0].Floats) == 0 {
+		return nil, nil
+	}
+
+	mergedSize := len(ds.data[0].Floats)
+	remainingSeriesWithFloats := 1
 	haveOverlaps := false
 
 	// We're going to create a new slice, so return this one to the pool.
 	// We'll return the other slices in the for loop below.
 	// We must defer here, rather than at the end, as the merge loop below reslices Floats.
 	// FIXME: this isn't correct for many-to-one / one-to-many matching - we'll need the series again (unless we store the result of the merge)
-	defer b.Pool.PutFPointSlice(data[0].Floats)
+	defer b.Pool.PutFPointSlice(ds.data[0].Floats)
```

Comment:
This could move beneath the `for`, and we could check if `remainingSeriesWithFloats` is 1 - if this is true, then we can just return that series without returning its `Floats` slice to the pool.